libraries/chain/include/eosio/chain/webassembly/common.hpp

```diff
@@ -104,6 +104,8 @@ namespace eosio { namespace chain {
       }
       EOS_VM_FROM_WASM(name, (uint64_t e)) { return name{e}; }
       uint64_t to_wasm(name&& n) { return n.to_uint64_t(); }
+      EOS_VM_FROM_WASM(float32_t, (float f)) { return ::to_softfloat32(f); }
+      EOS_VM_FROM_WASM(float64_t, (double f)) { return ::to_softfloat64(f); }
```

Comment:
I think this (and its conversion in the other direction which is missing here) should be moved into EOS VM (`as_value` and `as_result`). Also, `as_result` should be a lot more strict. It should only allow the return types: `int32_t`, `uint32_t`, `int64_t`, `uint64_t`, `float32_t`, `float64_t`, and `void*`. This will make it symmetric with what `as_value` allows (note that it would also need to be updated to take a `float32_t` rather than a `float` and a `float64_t` rather than a `double`).

Reply:
I don’t agree with this argument. Eos-vm is not eosio. I don’t want to enforce limitation on the base implementation unless it absolutely makes sense to do so. Most implementations that are going use eos-vm are not going to use softfloat (I would assume) and shouldn’t have the data types restricted at the eos-vm layer. Hardening of which types are useable is a per project decision.

Reply:
I don’t agree with this argument. Eos-vm is not eosio. I don’t want to enforce limitation on the base implementation unless it absolutely makes sense to do so. Most implementations that are going use eos-vm are not going to use softfloat (I would assume) and shouldn’t have the data types restricted at the eos-vm layer. Hardening of which types are useable is a per project decision.

Reply:
EOS VM still allows adding support for any return types desired by adding new `from_wasm` functions to a derived type converter class.

Requiring the user to harden the return types (rather than starting with a minimal set and allowing the user to expand as necessary) is an alternative approach. But that will require a lot more changes because preconditions do not have access to the return type.

Reply:
Intrinsics that return pointers and references are relatively rare.  They must point into linear memory, and can therefore only really return their arguments safely.

Using void* as the return type is not a significant burden:
  - Any intrinsic that returns a reference can be changed to return a pointer with some adjustments.
  - Any intrinsic that returns a pointer can trivially be changed to return [const] void*.
  - If using typed pointers or references as results is needed for some reason, they can easily be implemented in terms of void*.

Therefore, I do not believe that there is a strong reason to support T* and T& in the base implementation.

T* and T& are not supported for as_value.  For the sake of consistency, they should not be supported in as_result either.

In addition, for an extensible system, I generally favor keeping the base implementation minimal.

Overall, I think the balance is in favor of no support for T* and T&.